src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -507,23 +699,76 @@ protected StandardUsernameCredentials getCredentials() {
                                 GitClient.CREDENTIALS_MATCHER));
     }
 
-    protected abstract List<RefSpec> getRefSpecs();
+    /**
+     * @return the ref specs.
+     * @deprecated use {@link RefSpecsSCMSourceTrait}
+     */
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("3.4.0")
+    protected List<RefSpec> getRefSpecs() {
+        return Collections.emptyList();
+    }
+
+    /**
+     * Instantiates a new {@link GitSCMBuilder}.
+     * Subclasses should override this method if they want to use a custom {@link GitSCMBuilder} or if they need
+     * to pre-decorare the builder.
+     *
+     * @param head     the {@link SCMHead}.
+     * @param revision the {@link SCMRevision}.
+     * @return the {@link GitSCMBuilder}
+     * @see #decorate(GitSCMBuilder) for post-decoration.
+     */
+    protected GitSCMBuilder<?> newBuilder(@NonNull SCMHead head, @CheckForNull SCMRevision revision) {
+        return new GitSCMBuilder<>(head, revision, getRemote(), getCredentialsId());
+    }
+
+    /**
+     * Performs final decoration of the {@link GitSCMBuilder}. This method is called by
+     * {@link #build(SCMHead, SCMRevision)} immediately prior to returning {@link GitSCMBuilder#build()}.
+     * Subclasses should override this method if they need to overrule builder behaviours defined by traits.
+     *
+     * @param builder the builder to decorate.
+     */
+    protected void decorate(GitSCMBuilder<?> builder) {
+    }
 
+    /**
+     * {@inheritDoc}
+     */
     @NonNull
     @Override
     public SCM build(@NonNull SCMHead head, @CheckForNull SCMRevision revision) {
-        List<GitSCMExtension> extensions = new ArrayList<GitSCMExtension>(getExtensions());
-        if (revision instanceof SCMRevisionImpl) {
-            extensions.add(new BuildChooserSetting(new SpecificRevisionBuildChooser((SCMRevisionImpl) revision)));
+        GitSCMBuilder<?> builder = newBuilder(head, revision);
+        if (Util.isOverridden(AbstractGitSCMSource.class, getClass(), "getExtensions")) {
```

Comment:
@rsandell This method is the one that needs great care and attention as we discussed previously.

The assumption is that any new subclasses cannot override these methods as they are `@Restricted` thus there are either no traits (because old implementations cannot have overridden `getTraits()`) or there are only traits...

Hence these methods just apply the *equivalent* decoration to the builder to result in the same behaviour as before...

Now I have convinced myself of the equivalence, but I really need you to probe my self-convictions ;-)